File: .claude/skills/upstream-sync/refresh-patch-hashes.mjs
/**
 * Make pnpm-lock.yaml's patch hashes match the patch files on disk.
 *
 * pnpm records a sha256 of every file in `patchedDependencies` — and repeats it
 * inside every resolution key that depends on the patched package. A sync that
 * brings a changed patch leaves both stale, and `--frozen-lockfile` refuses the
 * tree. Recomputing beats re-resolving: resolution can fail for reasons
 * unrelated to the sync (a pin unpublished from the registry, a supply-chain
 * policy), and the dependency graph has not changed — only the bytes of a patch
 * the graph already references.
 *
 * The rewrite is `regenerate-xterm-patches.mjs`'s own, reused rather than
 * reimplemented: a first version here updated `patchedDependencies` alone and
 * left the resolution keys behind, which installs on a warm store and fails on
 * a cold one — which is CI.
 */
import { createHash } from 'node:crypto'
import { execFileSync } from 'node:child_process'
import { existsSync, readFileSync, writeFileSync } from 'node:fs'
import path from 'node:path'
import process from 'node:process'

import {
  lockfileHasPatchEntry,
  lockfilePatchHashIsStale,
  updateLockfilePatchHash
} from '../../../config/scripts/regenerate-xterm-patches.mjs'

const root = execFileSync('git', ['rev-parse', '--show-toplevel'], { encoding: 'utf8' }).trim()
const workspacePath = path.join(root, 'pnpm-workspace.yaml')
const lockPath = path.join(root, 'pnpm-lock.yaml')
if (!existsSync(workspacePath) || !existsSync(lockPath)) {
  process.exit(0)
}

const section = /^patchedDependencies:\n((?:[ \t]+.*\n)+)/m.exec(readFileSync(workspacePath, 'utf8'))
if (!section) {
  console.log('no patchedDependencies')
  process.exit(0)
}
const patches = [...section[1].matchAll(/^\s+'?([^':]+(?::[^']*)?)'?:\s*(\S+)\s*$/gm)].map(
  ([, key, file]) => [key.replace(/^'|'$/g, ''), file]
)

let lockfile = readFileSync(lockPath, 'utf8')
const refreshed = []
for (const [packageKey, relativePath] of patches) {
  const patchPath = path.join(root, relativePath)
  if (!existsSync(patchPath)) {
    console.log(`! ${packageKey}: ${relativePath} missing`)
    continue
  }
  if (!lockfileHasPatchEntry(lockfile, packageKey)) {
    continue
  }
  const hash = createHash('sha256').update(readFileSync(patchPath)).digest('hex')
  if (!lockfilePatchHashIsStale(lockfile, packageKey, hash)) {
    continue
  }
  lockfile = updateLockfilePatchHash(lockfile, packageKey, hash)
  refreshed.push(packageKey)
}
if (refreshed.length > 0) {
  writeFileSync(lockPath, lockfile)
  console.log(`refreshed ${refreshed.length}: ${refreshed.join(', ')}`)
} else {
  console.log('all patch hashes already match')
}

File: .claude/skills/upstream-sync/refresh-patch-hashes.py
#!/usr/bin/env python3
"""Make pnpm-lock.yaml's patch hashes match the patch files on disk.

pnpm records a sha256 of every file in `patchedDependencies`, so a sync that
brings a changed patch leaves the lockfile disagreeing with the tree and
`--frozen-lockfile` refuses it. Recomputing beats re-resolving: resolution can
fail for reasons unrelated to the sync (a pin unpublished from the registry, a
supply-chain policy), and the dependency graph has not changed — only the bytes
of a patch the graph already references.
"""
import hashlib
import pathlib
import re
import subprocess
import sys

root = pathlib.Path(subprocess.run(['git', 'rev-parse', '--show-toplevel'],
                                   capture_output=True, text=True).stdout.strip())
ws, lock = root / 'pnpm-workspace.yaml', root / 'pnpm-lock.yaml'
if not ws.exists() or not lock.exists():
    sys.exit(0)

section = re.search(r'^patchedDependencies:\n((?:  .*\n)+)', ws.read_text(), re.M)
if not section:
    print('no patchedDependencies'); sys.exit(0)
patches = dict(re.findall(r"^\s+'?([^':]+(?::[^']*)?)'?:\s*(\S+)\s*$", section.group(1), re.M))

text, changed = lock.read_text(), []
for dep, rel in patches.items():
    dep, path = dep.strip("'"), root / rel
    if not path.exists():
        print(f'! {dep}: {rel} missing'); continue
    digest = hashlib.sha256(path.read_bytes()).hexdigest()
    pattern = re.compile(rf"(^\s+'?{re.escape(dep)}'?:\s*)([0-9a-f]{{64}})\s*$", re.M)
    new_text, n = pattern.subn(lambda m: m.group(1) + digest, text)
    if n and new_text != text:
        changed.append(dep)
    text = new_text
if changed:
    lock.write_text(text)
    print(f'refreshed {len(changed)}: ' + ', '.join(changed))
else:
    print('all patch hashes already match')
